@InjectMock of a class with a method with generic throws type fails

### Describe the bug

When using the `@InjectMock` annotation on a class that has a method with 'throw' clause that uses a generic type, the test application fails to start, such as:

    public <X extends Exception> void useHandle() throws X

This pattern is [used by for instance the Jdbi library](https://github.com/jdbi/jdbi/blob/master/core/src/main/java/org/jdbi/v3/core/Jdbi.java#L409). The failure seems to be specific to the InjectMock flow, invoking `Mockito.mock` results in the expected behavior.

### Expected behavior

A mock is initialized and injected.

### Actual behavior

The following stack trace shows up:

```
java.lang.reflect.InvocationTargetException
        at java.base/java.lang.reflect.Method.invoke(Method.java:569)
        at io.quarkus.test.junit.AbstractTestWithCallbacksExtension.invokeCallbacks(AbstractTestWithCallbacksExtension.java:162)
        at io.quarkus.test.junit.AbstractTestWithCallbacksExtension.invokeBeforeEachCallbacks(AbstractTestWithCallbacksExtension.java:76)
        at io.quarkus.test.junit.QuarkusTestExtension.beforeEach(QuarkusTestExtension.java:398)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: java.lang.NoClassDefFoundError: X extends java/lang/Exception
        at java.base/java.lang.Class.getDeclaredMethods0(Native Method)
        at java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3402)
        at java.base/java.lang.Class.getDeclaredMethod(Class.java:2673)
        at io.quarkus.test.junit.MockSupport.installMock(MockSupport.java:53)
        at io.quarkus.test.junit.QuarkusMock.installMockForInstance(QuarkusMock.java:39)
        at io.quarkus.test.junit.mockito.internal.SetMockitoMockAsBeanMockCallback.installMock(SetMockitoMockAsBeanMockCallback.java:21)
        at java.base/java.lang.Iterable.forEach(Iterable.java:75)
        at io.quarkus.test.junit.mockito.internal.SetMockitoMockAsBeanMockCallback.beforeEach(SetMockitoMockAsBeanMockCallback.java:11)
        ... 6 more
Caused by: java.lang.ClassNotFoundException: X extends java.lang.Exception
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:576)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:523)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:576)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:523)
        ... 14 more
```

### How to Reproduce?

A simple reproduction is published here: https://github.com/yholkamp/quarkus-bug-repro
The test [GreetingResourceTest](https://github.com/yholkamp/quarkus-bug-repro/blob/main/src/test/java/org/acme/GreetingResourceTest.java#L12-L13) fails when requested mock injection of [MyTroublingClass](https://github.com/yholkamp/quarkus-bug-repro/blob/main/src/main/java/org/acme/MyTroublingClass.java)

### Output of `uname -a` or `ver`

Linux ydesk 6.11.0-26-generic #26~24.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Apr 17 19:20:47 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.15" 2025-04-15

### Quarkus version or git rev

3.23.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.9

### Additional information

_No response_

--- a/independent-projects/arc/processor/banned-signatures.txt
+++ b/independent-projects/arc/processor/banned-signatures.txt
@@ -1,0 +1,8 @@
+@defaultMessage Don't use Maven classes. They won't be available when using Gradle.
+org.apache.maven.**
+org.codehaus.plexus.**
+
+@defaultMessage Never use Type#toString() as it's almost always the wrong thing to do. Usually org.jboss.jandex.DotName#toString() is what is needed
+org.jboss.jandex.Type#toString()
+
+org.apache.commons.io.** @ Don't use commons-io dependency

--- a/independent-projects/arc/processor/pom.xml
+++ b/independent-projects/arc/processor/pom.xml
@@ -76,6 +76,9 @@
                                 <signaturesFile>./banned-signatures.txt</signaturesFile>
                             </signaturesFiles>
                             <ignoreSignaturesOfMissingClasses>true</ignoreSignaturesOfMissingClasses>
+                            <suppressAnnotations>
+                                <annotation>**.SuppressForbidden</annotation>
+                            </suppressAnnotations>
                         </configuration>
                         <phase>compile</phase>
                         <goals>

--- a/independent-projects/arc/processor/pom.xml
+++ b/independent-projects/arc/processor/pom.xml
@@ -63,5 +63,27 @@
         </dependency>
 
     </dependencies>
-
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>de.thetaphi</groupId>
+                <artifactId>forbiddenapis</artifactId>
+                <executions>
+                    <execution>
+                        <id>verify-forbidden-apis</id>
+                        <configuration>
+                            <signaturesFiles>
+                                <signaturesFile>./banned-signatures.txt</signaturesFile>
+                            </signaturesFiles>
+                            <ignoreSignaturesOfMissingClasses>true</ignoreSignaturesOfMissingClasses>
+                        </configuration>
+                        <phase>compile</phase>
+                        <goals>
+                            <goal>check</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
 </project>

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java
@@ -174,7 +174,7 @@ public class ClientProxyGenerator extends AbstractGenerator {
 
             // Exceptions
             for (Type exception : method.exceptions()) {
-                forward.addException(exception.toString());
+                forward.addException(exception.name().toString());
             }
             // Method params
             ResultHandle[] params = new ResultHandle[method.parametersCount()];

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ComponentsProviderGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ComponentsProviderGenerator.java
@@ -551,6 +551,7 @@ public class ComponentsProviderGenerator extends AbstractGenerator {
         }
 
         @Override
+        @SuppressForbidden(reason = "Using Type.toString() to build an informative message")
         void addComponentInternal(BeanInfo removedBean) {
 
             ResultHandle removedBeansHandle = addMethod.getMethodParam(0);
@@ -567,7 +568,7 @@ public class ComponentsProviderGenerator extends AbstractGenerator {
                 CatchBlockCreator catchBlock = tryBlock.addCatch(Throwable.class);
                 catchBlock.invokeStaticInterfaceMethod(
                         MethodDescriptors.COMPONENTS_PROVIDER_UNABLE_TO_LOAD_REMOVED_BEAN_TYPE,
-                        catchBlock.load(type.name().toString()), catchBlock.getCaughtException());
+                        catchBlock.load(type.toString()), catchBlock.getCaughtException());
                 AssignableResultHandle typeHandle = tryBlock.createVariable(Object.class);
                 try {
                     Types.getTypeHandle(typeHandle, tryBlock, type, tccl, typeCache);
@@ -654,13 +655,15 @@ public class ComponentsProviderGenerator extends AbstractGenerator {
         }
 
         @Override
+        @SuppressForbidden(reason = "Using Type.toString() to build an informative message")
         public ResultHandle get(org.jboss.jandex.Type type, BytecodeCreator bytecode) {
-            return bytecode.invokeInterfaceMethod(MethodDescriptors.MAP_GET, mapHandle, bytecode.load(type.name().toString()));
+            return bytecode.invokeInterfaceMethod(MethodDescriptors.MAP_GET, mapHandle, bytecode.load(type.toString()));
         }
 
         @Override
+        @SuppressForbidden(reason = "Using Type.toString() to build an informative message")
         public void put(org.jboss.jandex.Type type, ResultHandle value, BytecodeCreator bytecode) {
-            bytecode.invokeInterfaceMethod(MethodDescriptors.MAP_PUT, mapHandle, bytecode.load(type.name().toString()), value);
+            bytecode.invokeInterfaceMethod(MethodDescriptors.MAP_PUT, mapHandle, bytecode.load(type.toString()), value);
         }
 
     }

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ComponentsProviderGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ComponentsProviderGenerator.java
@@ -567,7 +567,7 @@ public class ComponentsProviderGenerator extends AbstractGenerator {
                 CatchBlockCreator catchBlock = tryBlock.addCatch(Throwable.class);
                 catchBlock.invokeStaticInterfaceMethod(
                         MethodDescriptors.COMPONENTS_PROVIDER_UNABLE_TO_LOAD_REMOVED_BEAN_TYPE,
-                        catchBlock.load(type.toString()), catchBlock.getCaughtException());
+                        catchBlock.load(type.name().toString()), catchBlock.getCaughtException());
                 AssignableResultHandle typeHandle = tryBlock.createVariable(Object.class);
                 try {
                     Types.getTypeHandle(typeHandle, tryBlock, type, tccl, typeCache);
@@ -655,12 +655,12 @@ public class ComponentsProviderGenerator extends AbstractGenerator {
 
         @Override
         public ResultHandle get(org.jboss.jandex.Type type, BytecodeCreator bytecode) {
-            return bytecode.invokeInterfaceMethod(MethodDescriptors.MAP_GET, mapHandle, bytecode.load(type.toString()));
+            return bytecode.invokeInterfaceMethod(MethodDescriptors.MAP_GET, mapHandle, bytecode.load(type.name().toString()));
         }
 
         @Override
         public void put(org.jboss.jandex.Type type, ResultHandle value, BytecodeCreator bytecode) {
-            bytecode.invokeInterfaceMethod(MethodDescriptors.MAP_PUT, mapHandle, bytecode.load(type.toString()), value);
+            bytecode.invokeInterfaceMethod(MethodDescriptors.MAP_PUT, mapHandle, bytecode.load(type.name().toString()), value);
         }
 
     }

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/InvokerGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/InvokerGenerator.java
@@ -791,13 +791,14 @@ public class InvokerGenerator extends AbstractGenerator {
             this.invoker = invoker;
         }
 
+        @SuppressForbidden(reason = "Using Type.toString() to build an informative message")
         CandidateMethod resolve() {
             if (matching.size() == 1) {
                 return matching.get(0);
             }
 
             if (matching.isEmpty()) {
-                String expectedType = this.expectedType.name().toString();
+                String expectedType = this.expectedType.toString();
                 String expectation = "";
                 if (transformer.isInputTransformer()) {
                     expectation = "\n"

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/InvokerGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/InvokerGenerator.java
@@ -797,7 +797,7 @@ public class InvokerGenerator extends AbstractGenerator {
             }
 
             if (matching.isEmpty()) {
-                String expectedType = this.expectedType.toString();
+                String expectedType = this.expectedType.name().toString();
                 String expectation = "";
                 if (transformer.isInputTransformer()) {
                     expectation = "\n"

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverGenerator.java
@@ -94,6 +94,7 @@ public class ObserverGenerator extends AbstractGenerator {
      *
      * @param observer
      */
+    @SuppressForbidden(reason = "Using Type.toString() to build an informative message")
     void precomputeGeneratedName(ObserverInfo observer) {
         // The name of the generated class differs:
         // "org.acme.Foo_Observer_fooMethod_hash" for normal observer where hash represents the signature of the observer method
@@ -117,7 +118,7 @@ public class ObserverGenerator extends AbstractGenerator {
             if (observer.getId() != null) {
                 sigBuilder.append(observer.getId());
             }
-            sigBuilder.append(observer.getObservedType().name().toString()).append(observer.getQualifiers().toString())
+            sigBuilder.append(observer.getObservedType().toString()).append(observer.getQualifiers().toString())
                     .append(observer.isAsync()).append(observer.getPriority()).append(observer.getTransactionPhase());
         } else {
             sigBuilder.append(observer.getObserverMethod().name())

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverGenerator.java
@@ -117,7 +117,7 @@ public class ObserverGenerator extends AbstractGenerator {
             if (observer.getId() != null) {
                 sigBuilder.append(observer.getId());
             }
-            sigBuilder.append(observer.getObservedType().toString()).append(observer.getQualifiers().toString())
+            sigBuilder.append(observer.getObservedType().name().toString()).append(observer.getQualifiers().toString())
                     .append(observer.isAsync()).append(observer.getPriority()).append(observer.getTransactionPhase());
         } else {
             sigBuilder.append(observer.getObserverMethod().name())

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java
@@ -714,7 +714,7 @@ public class SubclassGenerator extends AbstractGenerator {
             MethodCreator forward = delegateSubclass.getMethodCreator(methodDescriptor);
             // Exceptions
             for (Type exception : method.exceptions()) {
-                forward.addException(exception.toString());
+                forward.addException(exception.name().toString());
             }
 
             ResultHandle ret = null;

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SuppressForbidden.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SuppressForbidden.java
@@ -1,0 +1,15 @@
+package io.quarkus.arc.processor;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Used to suppress forbidden-apis errors inside the element it's added to.
+ */
+@Retention(RetentionPolicy.CLASS)
+@Target({ ElementType.CONSTRUCTOR, ElementType.FIELD, ElementType.METHOD, ElementType.TYPE })
+public @interface SuppressForbidden {
+    String reason();
+}

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/bcextensions/TypeImpl.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/bcextensions/TypeImpl.java
@@ -12,6 +12,8 @@ import jakarta.enterprise.lang.model.types.Type;
 
 import org.jboss.jandex.DotName;
 
+import io.quarkus.arc.processor.SuppressForbidden;
+
 abstract class TypeImpl<JandexType extends org.jboss.jandex.Type> extends AnnotationTargetImpl implements Type {
     final JandexType jandexType;
 
@@ -108,7 +110,8 @@ abstract class TypeImpl<JandexType extends org.jboss.jandex.Type> extends Annota
     }
 
     @Override
+    @SuppressForbidden(reason = "Using Type.toString() to build an informative message")
     public String toString() {
-        return jandexType.name().toString();
+        return jandexType.toString();
     }
 }

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/bcextensions/TypeImpl.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/bcextensions/TypeImpl.java
@@ -109,6 +109,6 @@ abstract class TypeImpl<JandexType extends org.jboss.jandex.Type> extends Annota
 
     @Override
     public String toString() {
-        return jandexType.toString();
+        return jandexType.name().toString();
     }
 }

